### ISSUE TITLE
H5 opener must default to meaningful chunk-size for unchunked datasets

### ISSUE THREAD

**Issue by axtimwalde** [saalfeldlab/bigcat#37]
https://github.com/saalfeldlab/bigcat/pull/32

introduces behavior that will render BigCAT non-functional in many everyday situations.  The dataset chunk-size of an HDF5 dataset is not necessarily meant to support interactive re-slicing, i.e. may have inconvenient dimensions, be too big or too small.  HDF5 loaders (unlike N5) do not require loading of blocks aligned to the dataset chunk-size.  I suggest the following approach:

1. get the dataset chunk size
2. calculate a meaningful default chunk-size using this approach https://github.com/saalfeldlab/bigcat/commit/2c21d125527ea94f07c205fdb5f824c62c1a8557
2. if the dataset is not chunked, us the meaningful default chunk size
3. if the dataset is chunked, compare the chunk-size to the meaningful default chunk size
4. if the difference in any dimension is > 100%, use the meaningful default chunk size
5. else use the dataset chunk size

**Comment by hanslovsky** [comment]
For some reason I had thought that hdf allows to compress without chunks, which was the rationale behind using the full image dimension when no chunks are specified. Clearly, that is not possible, though:
https://support.hdfgroup.org/HDF5/doc1.8/Advanced/Chunking/index.html

Regarding (5), do you mean `defaultChunkSize[ d ] > 2 * chunkSize[ d ]` for any `d`?

**Comment by axtimwalde** [comment]
```java
if (chunkSize[d] < defaultChunkSize[d] / 2 || chunkSize[d] > defaultChunkSize[d] * 2)
  chunkSize[d] = defaultChunkSize[d];
```

**Comment by hanslovsky** [comment]
I am currently working on this. In the progress of doing so, I was also cleaning up the backend dialog for h5 a little bit and succesfully switched to `N5Utils` and `n5-hdf5` (40b4eb6, this will make unifying n5 and hdf5 backend dialogs very easy). However, as far as I can tell, there is no way of loading data with a block size other than specified if it is specified in the h5 file:
https://github.com/saalfeldlab/n5-hdf5/blob/master/src/main/java/org/janelia/saalfeldlab/n5/hdf5/N5HDF5Reader.java#L296-L297
https://github.com/saalfeldlab/n5-imglib2/blob/07a09591a7b2462d8549877e2399e0f860fee594/src/main/java/org/janelia/saalfeldlab/n5/imglib2/N5Utils.java#L327
@axtimwalde Let me know if you know of any way around this. If not, we cannot use n5 for opening hdf5 files in this context.

**Comment by axtimwalde** [comment]
Not sure if this is what you mean:  If the dataset is not chunked, the default block size of the reader is used https://github.com/saalfeldlab/n5-hdf5/blob/master/src/main/java/org/janelia/saalfeldlab/n5/hdf5/N5HDF5Reader.java#L84-L95.  If the dataset is chunked, the chunk-size of the dataset is used.  This would be the same as with any other N5 dataset (except for the non-chunked case where you define the chunk-size).  Isn't that what would be expected?

**Comment by hanslovsky** [comment]
Yes, that is what would be expected with respect to n5. However, using a different chunk size than specified in the hdf5 file is not possible (items (4) and (5) in the initial comment), at least in my understanding:
  - if the dataset is chunked, compare the chunk-size to the meaningful default chunk size
  - if the difference in any dimension is > 100%, use the meaningful default chunk size

**Comment by axtimwalde** [comment]
Correct.  I will think about a solution (probably a flag to the reader that makes it override the blockSize even if chunk-size is given).  Should be optional though.

**Comment by hanslovsky** [comment]
Would it make sense to have two cached images? One that reads the h5 file, and then the second one that has appropriate block size and just copies data from the first image?

**Comment by axtimwalde** [comment]
No.

**Comment by hanslovsky** [comment]
Why not? Chunks must be loaded and decompressed completely anyway.

**Comment by axtimwalde** [comment]
If we cache the original chunks, we can also directly use them.  There
is no benefit in making another cache with the same data on top of
this, only wastes memory.  Or I am not seeing it.

**Comment by hanslovsky** [comment]
The benefit would be that we wouldn't need to load and decompress a single chunk multiple times. But maybe we should discuss in person with a whiteboard.

**Comment by hanslovsky** [comment]
After pen & paper discussion I agree that we should just manipulate the block size and not introduce another caching layer.

**Comment by hanslovsky** [comment]
Fixed in #83
